pulser/sampler/sampler.py

```diff
@@ -0,0 +1,279 @@
+
+    Returns:
+        A nested dictionnary of the samples of the amplitude, detuning and
+        phase at every nanoseconds for all channels.
+    """
+    if common_noises is None:
+        common_noises = []
+    if global_noises is None:
+        global_noises = []
+
+    # 1. determine if the global channel decay to a local one
+    # 2. extract samples
+    # 3. modulate
+    # 4. apply noises/SLM
+    # 5. write samples
+
+    samples: dict[str, list[QubitSamples]] = {}
+    addrs: dict[str, str] = {}
+
+    for ch_name, ch in seq.declared_channels.items():
+        s: list[QubitSamples]
+
+        addr = seq.declared_channels[ch_name].addressing
+
+        ch_noises = list(common_noises)
+
+        if addr == "Global":
+            decay = (
+                len(seq._slm_mask_targets) > 0
+                or len(global_noises) > 0
+                or len(common_noises) > 0
+            )
+            if decay:
+                addr = "Decayed"
+                ch_noises.extend(global_noises)
+
+        addrs[ch_name] = addr
+
+        strategy = _group_between_retargets if modulation else _regular
+        s = _sample_channel(seq, ch_name, strategy)
+        if modulation:
+            s = _modulate(ch, s)
+
+        s = noises.apply(s, ch_noises)
+
+        unmasked_qubits = seq._qids - seq._slm_mask_targets
+        s = [x for x in s if x.qubit in unmasked_qubits]  # SLM
+
+        samples[ch_name] = s
+
+    # format the samples in the simulation dict form
+    d = _write_dict(seq, samples, addrs)
+
+    return d
+
+
+def _prepare_dict(seq: Sequence, N: int) -> dict:
+    """Constructs empty dict of size N.
+
+    Usually N is the duration of seq.
+    """
+
+    def new_qty_dict() -> dict:
+        return {
+            "amp": np.zeros(N),
+            "det": np.zeros(N),
+            "phase": np.zeros(N),
+        }
+
+    def new_qdict() -> dict:
+        return {qubit: new_qty_dict() for qubit in seq._qids}
+
+    if seq._in_xy:
+        return {
+            "Global": {"XY": new_qty_dict()},
+            "Local": {"XY": new_qdict()},
+        }
+    else:
+        return {
+            "Global": defaultdict(new_qty_dict),
+            "Local": defaultdict(new_qdict),
+        }
+
+
+def _write_dict(
+    seq: Sequence,
+    samples: dict[str, list[QubitSamples]],
+    addrs: dict[str, str],
+) -> dict:
+    """Export the given samples to a nested dictionary."""
+    # Get the duration
+    if not _same_duration(samples):
+        raise ValueError("All the samples do not share the same duration.")
+    N = list(samples.values())[0][0].amp.size
+
+    d = _prepare_dict(seq, N)
+
+    for ch_name, some_samples in samples.items():
+        basis = seq.declared_channels[ch_name].basis
+        addr = addrs[ch_name]
+        if addr == "Global":
+            # Take samples on only one qubit and write them
+            a_qubit = next(iter(seq._qids))
+            to_write = [x for x in some_samples if x.qubit == a_qubit]
```

Comment:
Yes, that's what I do.
The reasoning behind this is that the sampler should be able to provide the info for each qubits independently. 
If the distinction between global and local channels is needed then we handle it at the output time.
It is also easier to handle per qubit noise on global channel by doing so.
 
Besides, this make the `sample()` function very straightforward since there is very little distinction between global and local channels.

For very large register it might becomes a bottle neck, but otherwise I do not think it worths being to clever about this ressource consumption.

Reply:
Yes, it's the large registers I'm worried about... It just seems very inefficient, but you may very well be right in assuming it doesn't cause any issues, I just don't know. 

Let's take for example a demanding case with 50 qubits, 3 channels, 10 µs of duration. If I'm not mistaken, this would amount to storing `50 * 3 * 10e3 * 3 = 4.5e6` floats, which amounts to 36 Mb. It seems manageable, so I won't push for cleverness here.

Reply:
I'll add a comment so that one day we may patch this if we happen to extract for too many qubits.